File: sections/21_graphs_i_mixed_contest/problems/f_component_repair/solution.py
import sys


def main():
    data = list(map(int, sys.stdin.buffer.read().split()))
    n, m = data[:2]
    graph = [[] for _ in range(n)]
    index = 2
    for _ in range(m):
        u, v = data[index] - 1, data[index + 1] - 1
        index += 2
        graph[u].append(v); graph[v].append(u)
    seen = [False] * n
    sizes = []
    for start in range(n):
        if seen[start]:
            continue
        seen[start] = True
        stack = [start]
        size = 0
        while stack:
            u = stack.pop()
            size += 1
            for v in graph[u]:
                if not seen[v]:
                    seen[v] = True
                    stack.append(v)
        sizes.append(size)
    print(len(sizes), len(sizes) - 1, max(sizes))


if __name__ == "__main__":
    main()

File: sections/21_graphs_i_mixed_contest/problems/f_component_repair/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n, m;
    cin >> n >> m;
    vector<vector<int>> graph(n);
    while (m--) {
        int u, v;
        cin >> u >> v;
        --u; --v;
        graph[u].push_back(v);
        graph[v].push_back(u);
    }
    vector<char> seen(n);
    int components = 0, largest = 0;
    for (int start = 0; start < n; ++start) if (!seen[start]) {
        ++components;
        int size = 0;
        vector<int> stack{start};
        seen[start] = true;
        while (!stack.empty()) {
            int u = stack.back();
            stack.pop_back();
            ++size;
            for (int v : graph[u]) if (!seen[v]) {
                seen[v] = true;
                stack.push_back(v);
            }
        }
        largest = max(largest, size);
    }
    cout << components << ' ' << components - 1 << ' ' << largest << '\n';
}
